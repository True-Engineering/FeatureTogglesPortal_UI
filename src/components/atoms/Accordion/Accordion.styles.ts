import {
  commonTheme,
  ComponentStyles,
} from '@true-engineering/true-react-common-ui-kit';

const { colors } = commonTheme;

export const styles = {
  root: {
    width: '100%',
  },

  item: {},

  header: {
    display: 'flex',
    alignItems: 'center',
    padding: [20, 14],
    backgroundColor: colors.CLASSIC_WHITE,
    cursor: 'pointer',

    '&:hover': {
      backgroundColor: colors.GREY_BACKGROUND,
    },
  },

  arrow: {
    width: 20,
    height: 20,
    transition: 'transform .2s ease',
    flexShrink: 0,
  },

  openArrow: {
    transform: 'rotate(90deg)',
  },

  title: {
    display: 'flex',
    flexGrow: 1,
    padding: [0, 6],
  },

  content: {
    display: 'flex',
    width: '100%',
  },
};

export type AccordionStyles = ComponentStyles<typeof styles>;
