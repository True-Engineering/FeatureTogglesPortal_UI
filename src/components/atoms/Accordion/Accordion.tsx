import React, { FC, useState } from 'react';
import { CSSTransition } from 'react-transition-group';
import clsx from 'clsx';
import { Icon } from '@true-engineering/true-react-common-ui-kit';
import { useFeatureTogglesTheme } from '../../../hooks';
import { useCollapseAnimationStyles } from '../../commonStyles';
import { AccordionStyles, styles } from './Accordion.styles';

export interface IAccordionItem {
  key: string | number;
  title: React.ReactNode;
  content: React.ReactNode;
  initialIsOpened?: boolean;
}
export interface IAccordionProps {
  items: IAccordionItem[];
  tweakStyles?: AccordionStyles;
}

const Accordion: FC<IAccordionProps> = ({ items, tweakStyles }) => {
  // чтобы классы не пересоздавались для каждого AccordionItem,
  // создадим стили здесь и передадим их через props
  const { classes } = useFeatureTogglesTheme('Accordion', styles, tweakStyles);

  return (
    <div className={classes.root}>
      {items.map(item => (
        <AccordionItem key={item.key} classes={classes} item={item} />
      ))}
    </div>
  );
};

export interface IAccordionItemProps {
  item: IAccordionItem;
  classes: Record<keyof AccordionStyles, string>;
}

const AccordionItem: FC<IAccordionItemProps> = ({
  item: { title, content, initialIsOpened = false },
  classes,
}) => {
  const collapseAnimationClasses = useCollapseAnimationStyles();
  const [isOpened, setIsOpened] = useState(initialIsOpened);

  return (
    <div className={classes.item}>
      <div className={classes.header} onClick={() => setIsOpened(v => !v)}>
        <div className={clsx(classes.arrow, isOpened && classes.openArrow)}>
          <Icon type="chevron-right" />
        </div>
        <div className={classes.title}>{title}</div>
      </div>

      <CSSTransition
        in={isOpened}
        timeout={300}
        unmountOnExit
        classNames={collapseAnimationClasses}
      >
        <div>
          <div className={classes.content}>{content}</div>
        </div>
      </CSSTransition>
    </div>
  );
};

export default Accordion;
