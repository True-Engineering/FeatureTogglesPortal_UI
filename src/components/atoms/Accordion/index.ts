export { default as Accordion } from './Accordion';
export * from './Accordion';

export type { AccordionStyles } from './Accordion.styles';
