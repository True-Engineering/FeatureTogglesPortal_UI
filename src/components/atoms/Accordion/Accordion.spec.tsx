import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent } from '@testing-library/react';
import Accordion from './Accordion';

const items = [
  {
    key: 1,
    title: 'Item 1',
    content: <div>Content 1</div>,
  },
  {
    key: 2,
    title: 'Item 2',
    content: <div>Content 2</div>,
  },
];

describe('Accordion корректно отрабатывает сценарии', () => {
  test('Рендер на мок данных', () => {
    const { container } = render(<Accordion items={items} />);

    expect(container).toMatchSnapshot();
  });

  test('Корректно отрабатывает клик', () => {
    const { getByText } = render(<Accordion items={items} />);

    fireEvent.click(getByText('Item 2'));
    expect(getByText('Content 2')).not.toBeNull();
  });
});
