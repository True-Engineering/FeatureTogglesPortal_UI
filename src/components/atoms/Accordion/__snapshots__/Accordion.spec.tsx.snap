// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion корректно отрабатывает сценарии Рендер на мок данных 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="item"
    >
      <div
        class="header"
      >
        <div
          class="arrow"
        >
          <div
            class="root"
          >
            <svg
              fill="none"
              height="100%"
              viewBox="0 0 20 20"
              width="100%"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M11.5858 10L7.29289 5.70711C6.90237 5.31658 6.90237 4.68342 7.29289 4.29289C7.68342 3.90237 8.31658 3.90237 8.70711 4.29289L13.7071 9.29289C14.0976 9.68342 14.0976 10.3166 13.7071 10.7071L8.70711 15.7071C8.31658 16.0976 7.68342 16.0976 7.29289 15.7071C6.90237 15.3166 6.90237 14.6834 7.29289 14.2929L11.5858 10Z"
                fill="currentColor"
                fill-rule="evenodd"
                stroke="none"
              />
            </svg>
          </div>
        </div>
        <div
          class="title"
        >
          Item 1
        </div>
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="header"
      >
        <div
          class="arrow"
        >
          <div
            class="root"
          >
            <svg
              fill="none"
              height="100%"
              viewBox="0 0 20 20"
              width="100%"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M11.5858 10L7.29289 5.70711C6.90237 5.31658 6.90237 4.68342 7.29289 4.29289C7.68342 3.90237 8.31658 3.90237 8.70711 4.29289L13.7071 9.29289C14.0976 9.68342 14.0976 10.3166 13.7071 10.7071L8.70711 15.7071C8.31658 16.0976 7.68342 16.0976 7.29289 15.7071C6.90237 15.3166 6.90237 14.6834 7.29289 14.2929L11.5858 10Z"
                fill="currentColor"
                fill-rule="evenodd"
                stroke="none"
              />
            </svg>
          </div>
        </div>
        <div
          class="title"
        >
          Item 2
        </div>
      </div>
    </div>
  </div>
</div>
`;
